sys/drv/bcm2835_emmc.c

```diff
@@ -0,0 +1,621 @@
+  resource_t *gpio = state->gpio;
+  int64_t r = 0;
+  /* GPIO_CD */
+  bcm2835_gpio_function_select(gpio, 47, BCM2835_GPIO_ALT3);
+  bcm2835_gpio_set_pull(gpio, 47, 2);
+  r = b_in(gpio, GPHEN1);
+  r |= 1 << 15;
+  b_out(gpio, GPHEN1, r);
+
+  /* GPIO_CLK, GPIO_CMD */
+  bcm2835_gpio_function_select(gpio, 48, BCM2835_GPIO_ALT3);
+  bcm2835_gpio_function_select(gpio, 49, BCM2835_GPIO_ALT3);
+  bcm2835_gpio_set_pull(gpio, 48, 2);
+  bcm2835_gpio_set_pull(gpio, 49, 2);
+
+  /* GPIO_DAT0, GPIO_DAT1, GPIO_DAT2, GPIO_DAT3 */
+  bcm2835_gpio_function_select(gpio, 50, BCM2835_GPIO_ALT3);
+  bcm2835_gpio_function_select(gpio, 51, BCM2835_GPIO_ALT3);
+  bcm2835_gpio_function_select(gpio, 52, BCM2835_GPIO_ALT3);
+  bcm2835_gpio_function_select(gpio, 53, BCM2835_GPIO_ALT3);
+  bcm2835_gpio_set_pull(gpio, 50, 2);
+  bcm2835_gpio_set_pull(gpio, 51, 2);
+  bcm2835_gpio_set_pull(gpio, 52, 2);
+  bcm2835_gpio_set_pull(gpio, 53, 2);
+}
+
+#define BCMEMMC_INIT_FREQ 400000
+
+static int bcmemmc_init(device_t *dev) {
+  assert(dev->driver == (driver_t *)&bcmemmc_driver);
+  bcmemmc_state_t *state = (bcmemmc_state_t *)dev->state;
+  resource_t *emmc = state->emmc;
+  int64_t r, cnt;
+
+  state->host_version =
+    (b_in(emmc, BCMEMMC_SLOTISR_VER) & HOST_SPEC_NUM) >> HOST_SPEC_NUM_SHIFT;
+  klog("e.MMC: GPIO set up");
+  /* Reset the card. */
+  b_out(emmc, BCMEMMC_CONTROL0, 0);
+  b_set(emmc, BCMEMMC_CONTROL1, C1_SRST_HC);
+  cnt = 10000;
+  do {
+    delay(30); /* ! */
+  } while ((b_in(emmc, BCMEMMC_CONTROL1) & C1_SRST_HC) && cnt--);
+  if (cnt <= 0) {
+    klog("ERROR: failed to reset EMMC");
+    return ETIMEDOUT;
+  }
+  klog("e.MMC: reset OK");
+  b_set(emmc, BCMEMMC_CONTROL1, C1_CLK_INTLEN | C1_TOUNIT_MAX);
+  /* Set clock to setup frequency. */
+  if ((r = bcmemmc_clk(dev, BCMEMMC_INIT_FREQ)))
+    return r;
+  b_out(emmc, BCMEMMC_INT_EN,
+        INT_CMD_DONE | INT_DATA_DONE | INT_READ_RDY | INT_WRITE_RDY |
+          INT_CMD_TIMEOUT | INT_DATA_TIMEOUT);
+  b_out(emmc, BCMEMMC_INT_MASK,
+        INT_CMD_DONE | INT_DATA_DONE | INT_READ_RDY | INT_WRITE_RDY |
+          INT_CMD_TIMEOUT | INT_DATA_TIMEOUT);
+
+  return 0;
+}
+
+static int bcmemmc_probe(device_t *dev) {
+  return dev->unit == 3;
+}
+
+DEVCLASS_DECLARE(emmc);
+
+static int bcmemmc_attach(device_t *dev) {
+  assert(dev->driver == (driver_t *)&bcmemmc_driver);
+  bcmemmc_state_t *state = (bcmemmc_state_t *)dev->state;
+
+  state->gpio = device_take_memory(dev, 0, RF_ACTIVE);
+  state->emmc = device_take_memory(dev, 1, RF_ACTIVE);
+
+  spin_init(&state->slock, 0);
+  cv_init(&state->cv_intr, "SD card response conditional variable");
+
+  b_out(state->emmc, BCMEMMC_INTERRUPT, 0xff);
+  emmc_gpio_init(dev);
+
+  state->irq = device_take_irq(dev, 2, RF_ACTIVE);
+  bus_intr_setup(dev, state->irq, bcmemmc_intr_filter, NULL, state,
+                 "e.MMMC interrupt");
+
+  int init_res = bcmemmc_init(dev);
+  if (init_res) {
+    klog("e.MMC initialzation failed with code %d.", init_res);
```

Comment:
I don't like this, but if you really wanna do this use `strerror()` to print the error code.

Reply:
No... `strerror` can't be used with `klog`. `klog` is lazy, it stores only pointer for string which will be invalidated with next `strerror` call. If you want to pass error in readable form to `klog` you need to allocate memory (for ever) and use `strerror_r`. I'm not sure if `strerror` is imported to kernel libraries.

Reply:
This is fine for debugging purposes and I'm pretty sure that `strerror` is not imported to libkern.